### DISCUSSION TITLE
Doubt around the Inception Score

### DISCUSSION THREAD

**Discussion by paudom** [Lightning-AI/torchmetrics#457] [Q&A] [answered]
Hi! After training a GAN and computing the Inception Score (IS) to evaluate the diversity, I'm confused with the result the metric gives.

If I understand correctly IS output should be between 1 and 1000 (because the inception network trained on ImageNet has 1000 classes) and a higher value means better quality/diversity. However, the value I obtain for the KL mean is very close to 1, while the KL std is almost 0. This confuses me because the images generated are pretty diverse from each other.

Here's an example:
![5 7 1](https://user-images.githubusercontent.com/37597137/129734348-60354d5f-d7af-4f8e-913e-24859e1dc4c4.png)

I would like to ask why the result is so close to 1. I've checked the documentation for the [IS metric](https://torchmetrics.readthedocs.io/en/latest/references/modules.html?highlight=IS#is) but the message under `compute` confuses me more:
_Override this method to compute the final metric value from state variables synchronized across the distributed backend._

Does this mean that I need to override the `compute` method to obtain a correct value for IS?

**Comment by SkafteNicki** [comment, marked as answer]
Hi,
So you should not override your own compute method. The reason that docstring is in the documenting is that it has not been overriden by the inception metric class:
https://github.com/PyTorchLightning/metrics/blob/00f0256db37406df09c18ddc18f4eca99ef7d59e/torchmetrics/image/inception.py#L158-L159
which means that it just defaults to the docstring of the metric class. 
https://github.com/PyTorchLightning/metrics/blob/00f0256db37406df09c18ddc18f4eca99ef7d59e/torchmetrics/metric.py#L375-L378
That said, I am not sure why you are getting a score around 1. If you could provide a file with a number of generated images we could maybe try to debug it for you?

**Reply by paudom** [reply]
I will try to upload a file with some generated images. 

By the way, the generated images are in the [-1, 1] range so in order to compute the Inception Score I first denormalize the images (range [0,1]) and then multiply by 255 and convert to torch.uint8 as the metric does not allow floats. 

There's something in this process that could be the reason for getting values close to 1?

**Reply by SkafteNicki** [reply]
It does not really sound like you have made something wrong there.
I am wondering if it is completely unreasonable that you get a score around 1. This would imply that basically all your images gets classified the same way by the inception net. I am not sure if that is unreasonable since it is the same object that is most in focus in all images: the bed.

**Reply by paudom** [reply]
This was overlooked by me and could very well be the reason. In fact, the GAN is only generating the same type of objects. Thanks a lot for the insight. 

I will try to upload the samples of images anyway, but your comment seems pretty reasonable. 

Thanks!

**Reply by paudom** [reply]
After looking at it more deeply I think you are on the right track. 

As mentioned in this [paper](https://arxiv.org/pdf/1706.02071.pdf), if the GAN is generating only one object, in order to evaluate the diversity across the same class category, it is necessary to compute the discrepancy across samples of the same category. The paper mentions that metric as modification of the Inception Score (m-IS)

A quick search for an implementation of this modified inception score could be this [solution](https://github.com/val-iisc/deligan/tree/master/src/modified_inception_scores).

**Reply by SkafteNicki** [reply]
Alright that sounds like a solution. You should be able to take the existing inception score metric in torchmetrics and just override the `compute` to be the modified version.
